### PR TITLE
feat(cdk): adding support ECMA Decorators for `@tuiPure`

### LINKED ISSUE #4641: 🐞 - @tuiPure do not work as ECMA decorator
### Playground Link

-

### Description

[ECMA decorators](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) have been added to typescript 5. @tuiPure is incompatible with it.

image

It is necessary to implement ECMA Decorators support with backward compatibility with Legacy Decorators.

In my project, the decorator implementation looks like this (maybe it can help you):

```
export const pure = <A extends unknown[], R>(
  target: (...args: A) => R,
  context: ClassMethodDecoratorContext | ClassGetterDecoratorContext,
) => {
  if (context.kind === 'method' || context.kind === 'getter') {
    let previousArgs: A | undefined;
    let result: R;
    return function (this: object, ...args: A): R {
      if (previousArgs?.length === args.length && previousArgs.every((item, index) => item === args[index])) {
        return result;
      }
      previousArgs = args;
      return (result = target.apply(this, args));
    };
  }
  return undefined;
};
```

### Angular version

16.1.0

### Taiga UI version

3.30.0

### Which browsers have you used?

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### PR DESCRIPTION
**Pull request by sviat9440**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #4641

### CODE REVIEW

**Comment by splincode** [comment]
@sviat9440 e2e are failed, is it expected?

**Comment by waterplea** [comment]
> @sviat9440 e2e are failed, is it expected?

Could be just a flaky test, I have rebased the brank, we will see how it goes the second time.

File: projects/cdk/decorators/pure.ts [hunk 1]
```diff
@@ -10,61 +61,65 @@ import {TuiPureException} from '@taiga-ui/cdk/exceptions';
...
     {get, enumerable, value}: TypedPropertyDescriptor<T>,
-): TypedPropertyDescriptor<T> {
+): TypedPropertyDescriptor<T>;
+
+/**
+ * Implements lazy initialization for getter or memoization of a function call similar to pure {@link: Pipe}.
+ * Replaces getter with its calculated value upon first call or keeps track of last call arguments and returned
+ * value for function, skipping calculation when arguments are strictly the same.
+ *
+ * CAUTION: `this` is not available inside such functions/getters, they must be pure.
```
[the remark this answers is not in the document]
Comment (sviat9440) [reply projects/cdk/decorators/pure.ts:74]: I just copied the description that was already there. Okay, I'll fix it.

**Review by waterplea — APPROVED** [review]

**Review by vladimirpotekhin — APPROVED** [review]

**Comment by splincode** [comment]
it is okay?
it is not flaky

image

**Comment by waterplea** [comment]
> it is okay? it is not flaky
> 
> image

I'll investigate it after the merge.

**Pull request merged**